gpii/node_modules/dbOperation/src/DbDataStoreUtils.js

```diff
@@ -0,0 +1,545 @@
+ * @param idName {String} The name for the unique id field. Usually "id".
+ * @param data {Object} The data to be saved in the new record
+ * @param idValue {String} Optional. The id value. If not provided, an UUID will be generated.
+ * @return {Promise} A promise for the save response
+*/
+gpii.dbOperation.dbDataStore.addRecord = function (dataSource, docType, idName, data, idValue) {
+    var promise = fluid.promise();
+
+    if (data !== undefined) {
+        var directModel = {};
+        directModel[idName] = idValue || uuid.v4();
+        fluid.extend(data, {type: docType});
+        var finalDirectModel = fluid.extend(true, {}, dataSource.options.directModel, directModel);
+        promise = dataSource.set(finalDirectModel, data);
+    } else {
+        fluid.log("dbOperation, addRecord(), error occurs: ", error);
+        var error = gpii.dbOperation.composeError(gpii.dbOperation.errors.missingDoc, {docName: docType});
+        promise.reject(error);
+    }
+    return promise;
+};
+
+/** Use the kettle dataSource `set` method to udpate a record by its id.
+ * @param dataSource {Component} An instance of gpii.dbOperation.dbDataSource that handles the record update
+ * @param docType {String} The expected document type in the updated record. See gpii.dbOperation.docTypes defined in
+ * %gpii-universal/gpii/node_modules/dbOperation/src/DbConst.js
+ * @param docId {String} The document id.
+ * @param data {Object} The data to be updated
+ * @return {Promise} A promise for the update response
+*/
+gpii.dbOperation.dbDataStore.updateRecord = function (dataSource, docType, docId, data) {
+    var promise = fluid.promise();
+    var error;
+
+    if (!data) {
+        error = gpii.dbOperation.composeError(gpii.dbOperation.errors.missingDoc, {docName: docType});
+        promise.reject(error);
+    } else if (data.type !== docType) {
+        error = gpii.dbOperation.composeError(gpii.dbOperation.errors.mismatchedDocType, {docType: docType});
+        promise.reject(error);
+    } else {
+        var directModel = {};
+        fluid.set(directModel, "id", docId);
+        var finalDirectModel = fluid.extend(true, {}, dataSource.options.directModel, directModel);
+        promise = dataSource.set(finalDirectModel, data);
+    }
+    return promise;
+};
+
+// General GPII keys Functions
+// ---------------------------
+
+/**
+ * Add a GPII key
+ * @param saveDataSource {Component} The saveDataSource component provided by gpii.dbOperation.dbDataStore
+ * @param gpiiKeyData {Object} The GPII key data. An example of the data:
+ *  {
+ *      gpiiKey: {String},  // Optional
+ *      prefsSafeId: {String},
+ *      prefsSetId: {String}
+ *  }
+ * If this parameter is not provided, a new GPII key is still generated but does not associate with a prefs safe.
+ * @return {Promise} A promise object that carries either a response returned from CouchDB/PouchDB for adding the
+ * GPII key record, or an error if `gpiiKeyData` parameter is not provided.
```

Comment:
Looks like when `gpiiKeyData` is not provided we default to

```
{
    prefsSafeId: null,
    prefsSetId: null
}
```

rather than raising an error.